src/plugins/intel_cpu/src/emitters/x64/jit_snippets_emitters.cpp

```diff
@@ -1351,69 +1351,63 @@ void HorizonMaxEmitter::emit_impl(const std::vector<size_t>& in,
+        perform_op(dst_zmm, aux_zmm);
+        h->vshuff32x4(aux_zmm, dst_zmm, dst_zmm, 0xB1);
+        perform_op(dst_zmm, aux_zmm);
+    } else if (isa == dnnl::impl::cpu::x64::avx2) {
+        Ymm dst_ymm = Ymm(out[0]);
+        Ymm aux_ymm = Ymm(aux_vec_idxs[0]);
+        h->vperm2i128(aux_ymm, dst_ymm, dst_ymm, 0x01);
+        perform_op(dst_ymm, aux_ymm);
     }
-    h->add(h->rsp, vlen);
+    h->uni_vshufps(aux_vmm, dst_vmm, dst_vmm, 0x4E);
+    perform_op(dst_vmm, aux_vmm);
+    h->uni_vshufps(aux_vmm, dst_vmm, dst_vmm, 0xB1);
+    perform_op(dst_vmm, aux_vmm);
 }
 
-HorizonSumEmitter::HorizonSumEmitter(dnnl::impl::cpu::x64::jit_generator* h, dnnl::impl::cpu::x64::cpu_isa_t isa, const std::shared_ptr<ov::Node>& n) :
-    jit_emitter(h, isa, n, Precision::FP32, emitter_in_out_map::vec_to_vec) {}
+void HorizonEmitter::perform_op(const Xbyak::Xmm &dst_xmm, const Xbyak::Xmm &src_xmm) const {
+    assert(!"Horizontal operation is not implemented.");
+}
 
-void HorizonSumEmitter::emit_impl(const std::vector<size_t>& in,
-                                  const std::vector<size_t>& out) const {
-    if (host_isa_ == dnnl::impl::cpu::x64::sse41) {
-        emit_isa<dnnl::impl::cpu::x64::sse41>(in, out);
-    } else if (host_isa_ == dnnl::impl::cpu::x64::avx2) {
-        emit_isa<dnnl::impl::cpu::x64::avx2>(in, out);
-    } else if (host_isa_ == dnnl::impl::cpu::x64::avx512_core) {
-        emit_isa<dnnl::impl::cpu::x64::avx512_core>(in, out);
-    } else {
-        IE_THROW() << "HorizonSum emitter doesn't support " << host_isa_;
-    }
+void HorizonEmitter::perform_op(const Xbyak::Ymm &dst_xmm, const Xbyak::Ymm &src_xmm) const {
+    assert(!"Horizontal operation is not implemented.");
 }
 
-template <dnnl::impl::cpu::x64::cpu_isa_t isa>
-void HorizonSumEmitter::emit_isa(const std::vector<size_t> &in, const std::vector<size_t> &out) const {
-    using Vmm = typename dnnl::impl::utils::conditional3<isa == dnnl::impl::cpu::x64::sse41,
-            Xmm, isa == dnnl::impl::cpu::x64::avx2, Ymm, Zmm>::type;
+HorizonMaxEmitter::HorizonMaxEmitter(dnnl::impl::cpu::x64::jit_generator* h, dnnl::impl::cpu::x64::cpu_isa_t isa, const std::shared_ptr<ov::Node>& n) :
+    HorizonEmitter(h, isa, n) {}
 
-    Vmm src_vmm = Vmm(in[0]);
-    Xmm dst_xmm = Xmm(out[0]);
-    Xmm aux_xmm = Xmm(aux_vec_idxs[0]);
-
-    Reg64 aux_reg = Reg64(aux_gpr_idxs[0]);
-
-    const size_t vlen = dnnl::impl::cpu::x64::cpu_isa_traits<isa>::vlen;
-    const size_t vec_size = vlen / sizeof(float);
-    h->sub(h->rsp, vlen);
-    h->uni_vmovups(h->ptr[h->rsp], src_vmm);
-    h->uni_vpxor(dst_xmm, dst_xmm, dst_xmm);
-    for (size_t i = 0; i < vec_size; i++) {
-        h->mov(aux_reg, h->ptr[h->rsp + i * sizeof(float)]);
-        h->vmovq(aux_xmm, aux_reg);
-        h->uni_vaddps(dst_xmm, dst_xmm, aux_xmm);
-    }
-    h->add(h->rsp, vlen);
+void HorizonMaxEmitter::perform_op(const Xbyak::Xmm &dst_xmm, const Xbyak::Xmm &src_xmm) const {
+    h->uni_vmaxps(dst_xmm, dst_xmm, src_xmm);
```

Comment:
We have the code duplication still 😄 

I'd suggest to create only one `HorizonEmitter`  with enum class of reduction ops `OpType = { sum, max }`. Constructor of `HorizonEmitter` init class field `OpType m_op_type`. The value depends on type of ngraph node shared pointer.
Then we have `HorizonEmitter::emit_isa(..)` that calls method `perform_op(..)`:
```
template<typename Vmm>
perform_op(const Vmm& a, const Vmm& b, const Vmm& c) {
   switch(m_op_type):
        - add
        - max
        - smth else
        - throw exception ('is not supported')
}
```

What do you think? If it's possible and better than the current way, let's do it. I just would like to avoid the code duplication 🙂

Reply:
I like this suggestion! Applied. Thanks Alexandra!